aya_tamimi/task_manager.py

```diff
@@ -0,0 +1,203 @@
+import utils
+from task import Task
+from urgentTask import UrgentTask
+import json
+
+
+tasks = []
+
+
+def menu():
+    print("\nTask Manager")
+    print("1. Add Task")
+    print("2. View Tasks")
+    print("3. Update Task")
+    print("4. Delete Task")
+    print("5. Search Task by status")
+    print("6. Exit")
+
+
+# This function add new task by taking user inputs  and validate each input so only valid data is accepted.
+# The task is then saved to either the regular task list or the urgent task list, depending on whether it is marked as urgent.
+def add_task():
+    # Loop until valid data is entered for the title
+    while True:
+        title = input("Enter your task title: ")
+        try:
+            utils.get_valid_input(title)
+            break # break only when validation function return True
+        except ValueError as e:
+            print(e) 
+
+    while True:
+        description = input("Enter task description: ")
+        try:
+            utils.get_valid_input(description)
+            break
+        except ValueError as e:
+            print(e)
+
+    while True:
+        due_date = input("Enter task due_date (in YYYY-MM-DD format): ")
+        try:
+            utils.validate_date(due_date)
+            break
+        except ValueError as e:
+            print(e)
+    while True:
+        status = input("Enter task status (InProgress , Completed): ").strip().lower()
+        try:
+            utils.validate_status(status)
+            break
+        except ValueError as e:
+            print(e)
+
+    is_urgent = input("is the task urgent(yes/no) ? ").strip().lower()
+    if is_urgent in ['yes' ,'y']:
+        while True:
+            priority = input("Enter task priority (High/Medium/Low): ").strip().lower()
+            try:
+                utils.validate_priority(priority)
+                break
+            except ValueError as e:
+                print(e)
+        task_instance = UrgentTask(priority ,title,description,due_date,status)
+        tasks.append(task_instance)
+        print("Urgent task successfully created.")
+        main()
+    else :
+        task_instance = Task(title,description,due_date,status)
+        tasks.append(task_instance)
+        print("task successfully created")
+        main()
+
+def view_tasks():
```

Comment:
For this function, I see two approaches:

- Good presentation but with longer execution time:
  I like how you print the tasks, but keep in mind that your code takes more time to execute due to the following:
  1. n time to get urgent_tasks
  2. n time to get regular_tasks
  3. m time to print urgent_tasks
  4. m time to print regular_tasks
  The total execution time is 2n + 2m.

- Bad presentation but faster execution:
   You can loop through the tasks once and print them directly, which would take n time.
   
Please don't change anything right now, but always consider the most efficient way to accomplish tasks. In this case, if performance is more important than presentation, we should go with the second option, and vice versa.

Reply:
Yes, I understand your point. I've taken steps to reduce the complexity by combining the two iterations into a single one.

Reply:
amazing